HPCC-25351 Add Add WsTopology.TpServiceTargetQuery

wangkx opened the pull request:
The ESP method is added to report the information about services
in a cloud environment.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

wangkx committed:
HPCC-25351 Add Add WsTopology.TpServiceTargetQuery

The ESP method is added to report the information about services
in a cloud environment.

Revise based on review
1. Only report public services
2. Remove the code for reporting the target attribute
3. Default the @tls to false

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

HPCCSmoketest commented:
Process of PR-14579, label: h25351 starts now.
The reason of this test is: Forced to re-test.
Commit ID: abdd95c97e799cd9ab8d16159dac12047a325af9
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-9.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-9.ca-central-1.compute.internal
Sha: abdd95c97e799cd9ab8d16159dac12047a325af9
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |146 |146 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |26 sec  (00:00:26) |
|setup (thor) |9 |9 |0 |42 sec  (00:00:42) |
|setup (roxie) |19 |19 |0 |17 sec  (00:00:17) |
|test (hthor) |954 |954 |0 |666 sec  (00:11:06) |
|test (thor) |865 |865 |0 |1138 sec  (00:18:58) |
|test (roxie) |1027 |1027 |0 |731 sec  (00:12:11) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  602 sec (00:10:02) |  108 sec (00:01:48) |   21 sec (00:00:21) |   19 sec (00:00:19) | 1448 sec (00:24:08) |   17 sec (00:00:17) | 2223 sec (00:37:03) |

wangkx commented:
@GordonSmith ECLWatch UI may use the changes in this PR.

jakesmith commented on esp/scm/ws_topology.ecm:
```diff
@@ -59,7 +59,17 @@ ESPStruct TpCluster
+    unsigned Port;
+    bool TLSSecure;
+    bool Public;
+    string Target;
```
what is target? not in any of the service type schema's/or in values.yaml afaics.

wangkx replied:
I saw the target from the roxie service:
services:
      - name: roxie
        type: roxie
        port: 9876
        target: roxie

jakesmith replied:
right ok, I see. The target is the roxie, the service name of this roxie is also named roxie.

jakesmith commented on esp/scm/ws_topology.ecm:
```diff
@@ -59,7 +59,17 @@ ESPStruct TpCluster
+    bool TLSSecure;
+    bool Public;
+    string Target;
+};
```
How is this info. going to be used?
In particular - if not public, then it is an internal service (used only internally).

I think need to 1st be sure what we want and how this/these services are going to be used.
If it's primarily to display on an eclwatch page for example, apart from *perhaps* ops. looking at the system, I don't think it's useful to display internal services.


Not much detail in the JIRA as to how this info. would be used. Can you expand ?

Out of interest, how does bare-metal expose the esp services running on other ports ?

wangkx replied:
A user may use the port: 8010, 8002, etc. Should a user need to know the TLS?
If we only want to expose public services, I can add the filter. Should we let Admin users know internal services?
In bare-metal mode, we display similar information on Operation/System Servers page.

jakesmith replied:
I suspect we don't want to show internal services by default, not sure in general.
If it's useful, can be acted upon, then it makes sense I think, if not, it either needs to be ignored, or shown separately somehow.
I'd just show the public services for now.

HPCCSmoketest commented:
Process of PR-14579, label: h25351 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 03bb2fe514da767e8fec5ad68c2e0feab4e91dbd
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-190.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-190.ca-central-1.compute.internal
Sha: 03bb2fe514da767e8fec5ad68c2e0feab4e91dbd
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |146 |146 |0 |0 |0 |50 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |22 sec  (00:00:22) |
|setup (thor) |9 |9 |0 |40 sec  (00:00:40) |
|setup (roxie) |19 |19 |0 |21 sec  (00:00:21) |
|test (hthor) |954 |954 |0 |639 sec  (00:10:39) |
|test (thor) |865 |865 |0 |1111 sec  (00:18:31) |
|test (roxie) |1027 |1027 |0 |700 sec  (00:11:40) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  583 sec (00:09:43) |  103 sec (00:01:43) |   19 sec (00:00:19) |   16 sec (00:00:16) | 1415 sec (00:23:35) |   17 sec (00:00:17) | 2161 sec (00:36:01) |

wangkx commented:
@jakesmith I have changed the code to only return public services. I also removed the target attribute since it only exists in roxie service and the roxie service is not a public service.

jakesmith commented:
>I have changed the code to only return public services. I also removed the target attribute since it only exists in roxie service and the roxie service is not a public service.

Actually, the roxie service is public.. It's the public (client) gateway to the roxie that's deployed.
NB: see kubectl get svc 

But that's effectively a bug in how services are generated, i.e. should generate public for this service too.
Please open a JIRA for that - and link it to this one.

jakesmith commented on esp/smc/SMCLib/TpWrapper.cpp:
```diff
@@ -1970,6 +1970,36 @@ void CTpWrapper::getAttPath(const char* Path,StringBuffer& returnStr)
+        serviceTarget->setName(name);
+        serviceTarget->setType(type);
+        serviceTarget->setPort(service.getPropInt("@port"));
+        if (service.hasProp("@tls")) //if not set, default to ?
```
I would remove the comment, I think it's fine to treat absence of this property as false.
So would prob. change to:
```
   if (service.getPropBool("@tls"))
```

wangkx replied:
Will change

jakesmith reviewed (commented):
@wangkx - looks fine, 1 minor comment

wangkx commented:
> But that's effectively a bug in how services are generated, i.e. should generate public for this service too. Please open a JIRA for that - and link it to this one.

https://track.hpccsystems.com/browse/HPCC-25382

HPCCSmoketest commented:
Process of PR-14579, label: h25351 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 05c35e108a50596701bdf4d728f458089557cedd
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-112.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-112.ca-central-1.compute.internal
Sha: 05c35e108a50596701bdf4d728f458089557cedd
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |146 |146 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |25 sec  (00:00:25) |
|setup (thor) |9 |9 |0 |40 sec  (00:00:40) |
|setup (roxie) |19 |19 |0 |21 sec  (00:00:21) |
|test (hthor) |959 |959 |0 |776 sec  (00:12:56) |
|test (thor) |870 |870 |0 |1118 sec  (00:18:38) |
|test (roxie) |1032 |1032 |0 |880 sec  (00:14:40) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  613 sec (00:10:13) |  116 sec (00:01:56) |   21 sec (00:00:21) |   19 sec (00:00:19) | 1430 sec (00:23:50) |   19 sec (00:00:19) | 2226 sec (00:37:06) |

jakesmith reviewed (commented):
@wangkx - looks good, please squash

HPCCSmoketest commented:
Process of PR-14579, label: h25351 starts now.
The reason of this test is: Forced to re-test.
Commit ID: c7f5319adaa05dc514256299eeca7bdf59a99b51
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-224.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-224.ca-central-1.compute.internal
Sha: c7f5319adaa05dc514256299eeca7bdf59a99b51
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |146 |146 |0 |0 |0 |53 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |23 sec  (00:00:23) |
|setup (thor) |9 |9 |0 |40 sec  (00:00:40) |
|setup (roxie) |19 |19 |0 |19 sec  (00:00:19) |
|test (hthor) |959 |959 |0 |804 sec  (00:13:24) |
|test (thor) |870 |870 |0 |1094 sec  (00:18:14) |
|test (roxie) |1032 |1032 |0 |870 sec  (00:14:30) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  593 sec (00:09:53) |  107 sec (00:01:47) |   21 sec (00:00:21) |   16 sec (00:00:16) | 1401 sec (00:23:21) |   19 sec (00:00:19) | 2165 sec (00:36:05) |

wangkx commented:
@jakesmith I did rebase and tests.

jakesmith commented:
@wangkx - looks good, thought of one more question though.

is getServiceTargets correct naming?

These are services, I think we're calling the main component queues "Targets".
So it may be better for these methods to be just called "getServices" without conflating with Targets.

Not sure. @ghalliday - what are your thoughts?

wangkx commented:
The existing WsTopology service already has a method named TpServiceQuery which reports the hpcc components (dali, esps, dfuservers, etc) in bare metal env. Should I merge the code there?

jakesmith commented:
> The existing WsTopology service already has a method named TpServiceQuery which reports the hpcc components (dali, esps, dfuservers, etc) in bare metal env. Should I merge the code there?

What is that used for/by ? It may not be very applicable to cloud, especially if esp services are now being returned from this new service.

wangkx commented:
> What is that used for/by ? It may not be very applicable to cloud, especially if esp services are now being returned from this new service.
For example, http://10.173.160.101:8010/#/stub/OPS-DL/SystemServersQuery-DL/SystemServersQueryWidgetIframeWidget
Mainly used by Operation team. The page shows Network Address, working folders, links to log files and configuration file. Gordon may hide it for clouds.

jakesmith commented:
> For example, http://10.173.160.101:8010/#/stub/OPS-DL/SystemServersQuery-DL/SystemServersQueryWidgetIframeWidget
> Mainly used by Operation team. The page shows Network Address, working folders, links to log files and configuration file. Gordon may hide it for clouds.

ok, then the 'services' in a cloud setup (i.e. those returned by this service and generated in the component config) are different.
And yes TpServiceQuery sounds like it should be suppressed in cloud.
I'm not sure it's just eclwatch though - what sense will this info. make in cloud world, if called by anything else?

If this (and other) services replace TpServiceQuery (and others), then they can be excluded from each build.

I suspect there's quite a few other things in WsTopology that don't quite make sense in cloud anymore, e.g. TpMachine/tpMachineType, or things with paths to logs, etc.
At some point it might make more sense to have a WsContainerServices for services calls like this one, leaving the bare-metal ones as they are.

wangkx commented:
> At some point it might make more sense to have a WsContainerServices for services calls like this one, leaving the bare-metal ones as they are.

Considering a possible future WsContainerServices (or a better name) service, it may be better not merge the code into the existing WsTopology.TpServiceQuery. We may give the TpServiceTargetQuery method a better name (TpContainerServiceQuery?) and, later on, move the whole method, and other methods (like the DropZone related methods) into the WsContainerServices service.

jakesmith commented:
> > At some point it might make more sense to have a WsContainerServices for services calls like this one, leaving the bare-metal ones as they are.
> 
> Considering a possible future WsContainerServices (or a better name) service, it may be better not merge the code into the existing WsTopology.TpServiceQuery. We may give the TpServiceTargetQuery method a better name (TpContainerServiceQuery?) and, later on, move the whole method, and other methods (like the DropZone related methods) into the WsContainerServices service.

It depends how much the existing WsTopology services and new container services overlap/diverge.
If it a lot of of WsTopology still makes sense, then it may make sense to be conjoined, trying to minimize the differences.
If they're quite different in many areas, and the containerized services are fewer, smaller and cleaner in a separate service, I suspect it is better to separate.

jakesmith commented:
@GordonSmith - can you take look at this from a functionality point of view.
This service calls currently returns a list of services that are public, i.e. open to the outside (there are other internal services that it does not show). NB: these are the 'services' generated into the global config.

What will eclwatch + other client want to see etc. ?
Is the info. per service adequate?

I think there's another JIRA re. a separate service call that will return the job queues that clients can submit to.

jakesmith reviewed (commented):
@wangkx - should this be closed / does https://github.com/hpcc-systems/HPCC-Platform/pull/14636 replace it?

wangkx commented:
Looks like this PR should be replaced by https://github.com/hpcc-systems/HPCC-Platform/pull/14636.
